### ISSUE TITLE
CopyableFailure returns strange things in python 2.5

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3031]
|<img alt="matejcik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| matejcik reported|
|-|-|
|Trac ID|trac#3031|
|Type|defect|
|Created|2008-02-12 18:13:35Z|
|Branch|https://github.com/twisted/twisted/tree/newstyle-exceptions-pb-3031|

due to exceptions being new-style classes in Python 2.5, the behavior of str(exceptiontype) has changed.
where it originally returned
```python
"module.TypeName"
```
now it returns
```python
"<class 'module.TypeName'>"
```

CopyableFailure sets its 'type' attribute to str(exceptiontype). Therefore, code that does its checks like
```python
if failure.type == 'module.TypeName':
    do_something()
```
will no longer work under Python 2.5

attached patch fixes this by setting 'type' to reflect.qual(exceptiontype), which returns "module.TypeName" in both Python versions

Attachments:

* [TwistedCore-2.5.0-CopyableFailure.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d4f/d4f9ea86999948c4060588e61484e8f36a06d7a2/8e8401f16fb01423f023d77423f3dd3a5e24bdd7.patch) (2745 bytes) - added by matejcik on 2008-02-12 18:48:16Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3031 3031
type__defect defect
reporter__matejcik matejcik
priority__highest highest
milestone__ 
branch__branches_newstyle_exceptions_pb_3031 branches/newstyle-exceptions-pb-3031
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1202840015000000 1202840015000000
changetime__1203265140000000 1203265140000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="matejcik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|matejcik commented|
|-|-|

awesome. now you can see the patch _and_ the unit test, with cool highlighting

(the trac is broken in an interesting way. but i don't want to play with that now)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [22553]) Branching to 'newstyle-exceptions-pb-3031'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [22554]) apply matejcik's patch

refs [#3031](https://github.com/twisted/twisted/issues/3031)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Made a few other minor changes, but the original patch seems basically correct to me.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

* pyflakes errors:
```
twisted/test/test_pbfailure.py:12: 'fail' imported but unused
```
 * please use assertIsInstance in the new tests
 * make the new test fit in 80 columns.

As it's fairly minimal, please merge once done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [22572]) remove unused import

refs [#3031](https://github.com/twisted/twisted/issues/3031)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [22573]) remove isinstance checks in the test, just use equality, wrap long lines

refs [#3031](https://github.com/twisted/twisted/issues/3031)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [22574]) Merge newstyle-exceptions-pb-3031

Author: matejcik, exarkun
Reviewer: therve
Fixes [#3031](https://github.com/twisted/twisted/issues/3031)

Add tests for the behavior of PB remote methods which fail with unjellyable
exceptions.  Fix the implementation so that the type of the resulting
CopyableFailure is (once again) the fully-qualified Python name of the
exception which was raised.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
